File: archive/reduck/claude_client.py
"""Streaming interface to Claude Code via Agent SDK."""

import logging
import os
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass, replace
from pathlib import Path
from typing import Literal, cast

from claude_agent_sdk import (
    AssistantMessage,
    ClaudeAgentOptions,
    ResultMessage,
    ToolResultBlock,
    ToolUseBlock,
    UserMessage,
    query,
)
from claude_agent_sdk.types import StreamEvent

from reduck.models import path_to_slug

type PermissionMode = Literal["default", "acceptEdits", "plan", "bypassPermissions"]

log = logging.getLogger("claude")

# Prevent nested session error when running inside Claude Code
_ = os.environ.pop("CLAUDECODE", None)


# ── Config ──────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class ClaudeConfig:
    """Claude Code environment config.

    Two fields, one method. Built from env vars at server startup.

    Config loading order (later wins):
      1. Defaults (~/.claude, claude on PATH)
      2. .env file in cwd (loaded via python-dotenv)
      3. Actual env vars (override .env)
    """

    config_dir: str = "~/.claude"
    cli_path: str | None = None  # None = `claude` on PATH

    def subprocess_env(self) -> dict[str, str]:
        """Build env dict for the SDK subprocess."""
        env = {k: v for k, v in os.environ.items() if k != "CLAUDECODE"}
        expanded = os.path.expanduser(self.config_dir)
        if self.config_dir != "~/.claude":
            env["CLAUDE_CONFIG_DIR"] = expanded
        return env

    def project_dir(self, cwd: str) -> Path:
        """Session directory for a given project cwd."""
        return (
            Path(os.path.expanduser(self.config_dir)) / "projects" / path_to_slug(cwd)
        )


# ── Chunk types ─────────────────────────────────────────────────────────────


@dataclass
class TextDelta:
    text: str


@dataclass
class ContentBlockChunk:
    block: dict[str, object]


@dataclass
class Result:
    session_id: str
    cost_usd: float | None
    duration_ms: int
    error: str | None = None


type Chunk = TextDelta | ContentBlockChunk | Result


# ── Client ──────────────────────────────────────────────────────────────────


class Claude:
    """Streaming Claude Code interface.

    Stateless — holds config only. cwd is per-call (different sessions
    can target different projects).
    """

    _config: ClaudeConfig
    _stderr: Callable[[str], None]

    def __init__(self, config: ClaudeConfig):
        self._config = config
        self._stderr = lambda line: log.debug("sdk: %s", line.rstrip())

    async def converse(
        self,
        message: str,
        model: str,
        system_prompt: str,
        cwd: str,
        session_id: str | None = None,
        permission_mode: PermissionMode = "plan",
        fork: bool = False,
    ) -> AsyncIterator[Chunk]:
        options = ClaudeAgentOptions(
            model=model,
            cwd=cwd,
            system_prompt=system_prompt,
            include_partial_messages=True,
            permission_mode=permission_mode,
            allowed_tools=["Read", "WebSearch"],
            disallowed_tools=["AskUserQuestion", "Skill"],
            env=self._config.subprocess_env(),
            stderr=self._stderr,
        )
        if self._config.cli_path:
            options = replace(
                options, cli_path=os.path.expanduser(self._config.cli_path)
            )
        if session_id:
            options = replace(options, resume=session_id, fork_session=fork)
            log.info("resuming session %s (fork=%s)", session_id, fork)

        log.info("query: %s", message[:120])
        async for msg in query(prompt=message, options=options):
            if isinstance(msg, StreamEvent):
                delta = cast(dict[str, str], msg.event.get("delta", {}))
                if text := delta.get("text"):
                    yield TextDelta(text=text)
            elif isinstance(msg, AssistantMessage):
                for block in msg.content:
                    if isinstance(block, ToolUseBlock):
                        yield ContentBlockChunk(
                            block={
                                "type": "tool_use",
                                "id": block.id,
                                "name": block.name,
                                "input": block.input,
                            }
                        )
            elif isinstance(msg, UserMessage):
                for block in msg.content if isinstance(msg.content, list) else []:
                    if isinstance(block, ToolResultBlock):
                        raw = block.content
                        yield ContentBlockChunk(
                            block={
                                "type": "tool_result",
                                "tool_use_id": block.tool_use_id,
                                "content": raw
                                if isinstance(raw, str)
                                else str(raw)
                                if raw
                                else "",
                            }
                        )
            elif isinstance(msg, ResultMessage):
                error = msg.result if msg.is_error else None
                log.info(
                    "result: session=%s, cost=$%s, %dms, error=%s",
                    msg.session_id,
                    msg.total_cost_usd,
                    msg.duration_ms,
                    error,
                )
                yield Result(
                    session_id=msg.session_id,
                    cost_usd=msg.total_cost_usd,
                    duration_ms=msg.duration_ms,
                    error=error,
                )

File: src/server/claude-client.ts
/**
 * Streaming interface to Claude Code via the TS Agent SDK.
 * Port of Python claude_client.py.
 */

import { join } from 'node:path';
import { query } from '@anthropic-ai/claude-agent-sdk';
import type {
  Options,
  PermissionMode,
  SDKMessage,
  SDKPartialAssistantMessage,
  SDKAssistantMessage,
  SDKUserMessage,
  SDKResultMessage,
} from '@anthropic-ai/claude-agent-sdk';
import { pathToSlug } from '../shared/models.js';

// Prevent nested session error when running inside Claude Code
delete process.env['CLAUDECODE'];

// --- Config ---

export interface ClaudeConfig {
  configDir: string;     // e.g. "~/.claude"
  cliPath?: string;      // None = `claude` on PATH
}

export function subprocessEnv(config: ClaudeConfig): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [k, v] of Object.entries(process.env)) {
    if (k !== 'CLAUDECODE' && v !== undefined) {
      env[k] = v;
    }
  }
  const expanded = config.configDir.replace(/^~/, process.env['HOME'] || '');
  if (config.configDir !== '~/.claude') {
    env['CLAUDE_CONFIG_DIR'] = expanded;
  }
  return env;
}

export function projectDir(config: ClaudeConfig, cwd: string): string {
  const expanded = config.configDir.replace(/^~/, process.env['HOME'] || '');
  return join(expanded, 'projects', pathToSlug(cwd));
}

// --- Chunk types ---

export interface TextDelta {
  kind: 'text';
  text: string;
}

export interface ContentBlockChunk {
  kind: 'block';
  block: Record<string, unknown>;
}

export interface Result {
  kind: 'result';
  sessionId: string;
  costUsd: number | null;
  durationMs: number;
  error: string | null;
}

export type Chunk = TextDelta | ContentBlockChunk | Result;

// --- Client ---

export class Claude {
  private readonly config: ClaudeConfig;

  constructor(config: ClaudeConfig) {
    this.config = config;
  }

  async *converse(
    message: string,
    opts: {
      model: string;
      systemPrompt: string;
      cwd: string;
      sessionId?: string;
      permissionMode?: PermissionMode;
      fork?: boolean;
    },
  ): AsyncGenerator<Chunk> {
    const options: Options = {
      model: opts.model,
      cwd: opts.cwd,
      systemPrompt: opts.systemPrompt,
      includePartialMessages: true,
      permissionMode: opts.permissionMode ?? 'plan',
      allowedTools: ['Read', 'WebSearch'],
      disallowedTools: ['AskUserQuestion', 'Skill'],
      env: subprocessEnv(this.config),
      stderr: (line: string) => console.debug('sdk:', line.trimEnd()),
    };

    if (this.config.cliPath) {
      options.pathToClaudeCodeExecutable = this.config.cliPath.replace(
        /^~/,
        process.env['HOME'] || '',
      );
    }

    if (opts.sessionId) {
      options.resume = opts.sessionId;
      options.forkSession = opts.fork ?? false;
      console.info(`resuming session ${opts.sessionId} (fork=${opts.fork ?? false})`);
    }

    console.info(`query: ${message.slice(0, 120)}`);
    const stream = query({ prompt: message, options });

    for await (const msg of stream as AsyncIterable<SDKMessage>) {
      if (msg.type === 'stream_event') {
        const partial = msg as SDKPartialAssistantMessage;
        const event = partial.event as unknown as Record<string, unknown>;
        const delta = event['delta'] as Record<string, unknown> | undefined;
        if (delta) {
          const text = delta['text'];
          if (typeof text === 'string' && text) {
            yield { kind: 'text', text };
          }
        }
      } else if (msg.type === 'assistant') {
        const asst = msg as SDKAssistantMessage;
        const content = asst.message?.content;
        if (Array.isArray(content)) {
          for (const block of content) {
            const b = block as unknown as Record<string, unknown>;
            if (b['type'] === 'tool_use') {
              yield {
                kind: 'block',
                block: {
                  type: 'tool_use',
                  id: b['id'],
                  name: b['name'],
                  input: b['input'],
                },
              };
            }
          }
        }
      } else if (msg.type === 'user') {
        const user = msg as SDKUserMessage;
        const content = user.message?.content;
        if (Array.isArray(content)) {
          for (const block of content) {
            const b = block as unknown as Record<string, unknown>;
            if (b['type'] === 'tool_result') {
              const raw = b['content'];
              yield {
                kind: 'block',
                block: {
                  type: 'tool_result',
                  tool_use_id: b['tool_use_id'],
                  content: typeof raw === 'string' ? raw : raw ? String(raw) : '',
                },
              };
            }
          }
        }
      } else if (msg.type === 'result') {
        const result = msg as SDKResultMessage;
        // Error handling: success has `result`, error subtypes have `errors[]`
        let error: string | null = null;
        if (result.is_error) {
          if ('errors' in result && Array.isArray(result.errors)) {
            error = result.errors.join('; ');
          } else if ('result' in result) {
            error = String(result.result);
          }
        }
        console.info(
          `result: session=${result.session_id}, cost=$${result.total_cost_usd}, ${result.duration_ms}ms, error=${error}`,
        );
        yield {
          kind: 'result',
          sessionId: result.session_id,
          costUsd: result.total_cost_usd,
          durationMs: result.duration_ms,
          error,
        };
      }
      // Ignore other message types (system init, tool_progress, etc.)
    }
  }
}
